Signal/src/ViewControllers/HomeView/ConversationSearchViewController.swift

```diff
@@ -0,0 +1,244 @@
+
+    var uiDatabaseConnection: YapDatabaseConnection {
+        // TODO do we want to respond to YapDBModified? Might be hard when there's lots of search results, for only marginal value
+        return OWSPrimaryStorage.shared().uiDatabaseConnection
+    }
+
+    var searcher: ConversationSearcher {
+        return ConversationSearcher.shared
+    }
+
+    enum SearchSection: Int {
+        case conversations = 0
+        case contacts = 1
+        case messages = 2
+    }
+
+    // MARK: View Lifecyle
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        tableView.rowHeight = UITableViewAutomaticDimension
+        tableView.estimatedRowHeight = 60
+
+        tableView.register(ChatSearchResultCell.self, forCellReuseIdentifier: ChatSearchResultCell.reuseIdentifier)
+        tableView.register(MessageSearchResultCell.self, forCellReuseIdentifier: MessageSearchResultCell.reuseIdentifier)
+    }
+
+    // MARK: UITableViewDataSource
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        guard let searchSection = SearchSection(rawValue: section) else {
+            owsFail("unknown section: \(section)")
+            return 0
+        }
+
+        switch searchSection {
+        case .conversations:
+            return searchResultSet.conversations.count
+        case .contacts:
+            return searchResultSet.contacts.count
+        case .messages:
+            return searchResultSet.messages.count
+        }
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+
+        guard let searchSection = SearchSection(rawValue: indexPath.section) else {
+            return UITableViewCell()
+        }
+
+        switch searchSection {
+        case .conversations:
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: ChatSearchResultCell.reuseIdentifier) as? ChatSearchResultCell else {
+                return UITableViewCell()
+            }
+
+            guard let searchResult = self.searchResultSet.conversations[safe: indexPath.row] else {
+                return UITableViewCell()
+            }
+            cell.configure(searchResult: searchResult)
+            return cell
+        case .contacts:
+            // TODO
+            return UITableViewCell()
+        case .messages:
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: MessageSearchResultCell.reuseIdentifier) as? MessageSearchResultCell else {
+                return UITableViewCell()
+            }
+
+            guard let searchResult = self.searchResultSet.messages[safe: indexPath.row] else {
+                return UITableViewCell()
+            }
+
+            cell.configure(searchResult: searchResult)
+            return cell
+        }
+    }
+
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        return 3
```

Comment:
return `SearchSection.allCases.count`

Reply:
I didn't know about `allCases` - looks like it's slated for Swift 4.2.

https://github.com/apple/swift-evolution/blob/master/proposals/0194-derived-collection-of-enum-cases.md